api/apps/api/test/access-control/project-acl-get.e2e-spec.ts

```diff
@@ -0,0 +1,132 @@
+import { FixtureType } from '@marxan/utils/tests/fixture-type';
+import { getFixtures } from './project-acl.fixtures';
+
+let fixtures: FixtureType<typeof getFixtures>;
+
+beforeEach(async () => {
+  fixtures = await getFixtures();
+});
+afterEach(async () => {
+  await fixtures?.cleanup();
+});
+
+test(`getting project users when user is not added to project`, async () => {
+  const projectId = await fixtures.GivenProjectWasCreated();
+  const response = await fixtures.WhenGettingProjectUserAsNotInProject(
+    projectId,
+  );
+  fixtures.ThenForbiddenIsReturned(response);
+});
+
+test(`getting project users as the owner and unique user`, async () => {
+  const projectId = await fixtures.GivenProjectWasCreated();
+  const response = await fixtures.WhenGettingProjectUsersAsOwner(projectId);
+  fixtures.ThenSingleOwnerUserInProjectIsReturned(response);
+});
+
+test(`getting project users with user in project but is viewer`, async () => {
+  const projectId = await fixtures.GivenProjectWasCreated();
+  await fixtures.WhenAddingANewViewerToTheProjectAsOwner(projectId);
+  const response = await fixtures.WhenGettingProjectUsersAsViewer(projectId);
+  fixtures.ThenForbiddenIsReturned(response);
+});
+
+test(`getting project users with user in project but is contributor`, async () => {
+  const projectId = await fixtures.GivenProjectWasCreated();
+  await fixtures.WhenAddingANewContributorToTheProjectAsOwner(projectId);
+  const response = await fixtures.WhenGettingProjectUsersAsContributor(
+    projectId,
+  );
```

Comment:
Usually there should be only one "action", the other things should be either conditions (Given) or checks (Then). As in this scope we check the ACLs, `WhenGettingProjectUsersAsContributor` could be merged with `Then` (and be like check of particular permission).

Also the description isn't clear to me. Did you mean, `getting project users as contributor` ? (also even if so, test seems to be wrong 🤔 )

Reply:
The intention with all this tests was to be able to check that no other users were able to access the users of a given project except those with `project_owner` role.

I think this convoluted way of doing the tests (not only this one but the others as well) was because I didn't find an easy way of creating new users from the beginning, and for some of the tests I had some problems if I tried to add them in the `beforeEach()` statement.

I just noted down the general idea on how to improve them (splitting them up into revoke-get-add, less actions in the middle) and  added it to a TO-DO-FUTURE-PR note.

For the time being, and as we discussed earlier, I think it would be better if I try to finish all the tasks related to project-users access-control to use it as kind of chaotic approach to all this (as long as there is not a big security issue, or something directly wrong)

Reply:
dealing with ACL mechanics in tests is often tricky, and often things such as users need to be treated differently than in tests for other parts of the domain - what you have done here doesn't feel necessarily convoluted to me; nevertheless the test does seem off to me too: I am concerned that with most things user-related going on "behind the scenes", we can accidentally mess things up.

Partly matter of preference, but even just returning the new contributor user here and then passing it in to the then-check (which should then be named differently) would help. Or skipping the second condition (creation of a user) and encoding that in the then-check.

Pragmatically, I am also happy to keep these tests as they are now, since it's a tricky part and trying to get absolutely everything "right" would be too much for a single PR, in my opinion. But in order for us to keep confidence in these access-related tests over time, I think some refactor would be advisable in a future PR.

Reply:
actually, having looked now at the fixtures file, I would go directly for the second approach I suggested above (which should be the same suggested by Kamil earlier), so
given project was created
when getting users as contributor
then...

the fact that a contributor user was created at some point doesn't have much to do with what we are testing in this specific test, I think

Reply:
and so on for the rest of the tests and related fixtures - I'm not going to comment further given that they stick to similar patterns throughout.